.github/actions/setup-build/action.yml

```diff
@@ -0,0 +1,120 @@
+      name: (Linux,macOS) Platform config
+      run: |
+        echo "CABAL_PKGS_DIR=~/.cabal/packages" >> $GITHUB_ENV
+      shell: bash
+
+    # Needs to be before Cache Cabal so the cache can detect changes to the modified cabal.project file
+    - if: inputs.ghc == '9.0.1' || inputs.ghc == '9.2.1'
+      name: (GHC 9.0/9.2) Use modified `cabal.project`
+      env:
+        GHCVER: ${{ inputs.ghc }}
+      run: |
+        # File has some protections preventing regular `rm`.
+        # (most probably sticky bit is set on $HOME)
+        # `&&` insures `rm -f` return is positive.
+        # Many platforms aslo have `alias cp='cp -i'`.
+        rm -f -v cabal.project && cp -v cabal-ghc${GHCVER//./}.project cabal.project
+      shell: bash
+
+    - if: inputs.os == 'Windows' && inputs.ghc == '8.8.4'
+      name: (Windows,GHC 8.8) Modify `cabal.project` to workaround segfaults
+      run: |
+        echo "package floskell" >> cabal.project
+        echo "  ghc-options: -O0" >> cabal.project
+      shell: bash
+
+    # Shorten binary names as a workaround for filepath length limits in Windows,
+    # but since tests are hardcoded on this workaround -
+    # all platforms (in 2021-12-07) need it.
+    # All workflows which distinquishes cache on `cabal.project` needs this.
+    - name: Workaround shorten binary names
+      run: |
+        sed -i.bak -e 's/haskell-language-server/hls/g' \
+                    -e 's/haskell_language_server/hls/g' \
+                    haskell-language-server.cabal cabal.project
+        sed -i.bak -e 's/Paths_haskell_language_server/Paths_hls/g' \
+                    src/**/*.hs exe/*.hs
+      shell: bash
+
+    - name: Retrieving `cabal.project` Hackage timestamp
+      run: |
+        # Form: index-state: 2021-11-29T08:11:08Z
+        INDEX_STATE_ENTRY=$(grep index-state cabal.project)
+        # Form: 2021-11-29T08-11-08Z
+        INDEX_STATE1=$(echo "$INDEX_STATE_ENTRY" | cut -d' ' -f2 | tr ':' '-')
+        echo "INDEX_STATE=$INDEX_STATE1" >> $GITHUB_ENV
+      shell: bash
+
+    # We have to restore package sources before `cabal update`
+    # cause it overwrites the hackage index with the cached one
+    - name: Hackage sources cache
+      uses: actions/cache@v2
+      env:
+        cache-name: hackage-sources
+      with:
+        path:  ${{ env.CABAL_PKGS_DIR }}
+        key:   ${{ env.cache-name }}-${{ env.INDEX_STATE }}
+        restore-keys: ${{ env.cache-name }}-
+
+    # To ensure we get the lastest hackage index without telying in the haskell action logic
+    # It has to be done before `cabal freeze` to make it aware of the new index
+    - run: cabal update
+      shell: bash
+
+    - name: Form the package list ('cabal.project.freeze')
+      run: |
+        cabal v2-freeze && \
+        echo "" && \
+        echo 'Output:' && \
+        echo "" && \
+        cat 'cabal.project.freeze' && \
+        echo "" || \
+        echo 'WARNING: Could not produce the `freeze`.'
```

Comment:
Hm, is there maybe a way to set `-e` globally so we can just fail if anything fails?

Reply:
afaik the shell has -e by default but we want the opposite thing: it should not fail so cache will fallback to only index state

Reply:
But why? We can fail to produce the freeze file? Probably could do with a note explaining that if it's expected...

Reply:
good question, we had some issues producing it for bleeding-edge branches like ghc-9.2 but i *think* it should work now. However i am not sure if we have to be stricter and block ci. Maybe we could try the strict way and make it optional i we see there are more problems hard to fix quickly

Reply:
SGTM.

Also, further dumb question: why do we need the freeze file? Given that we pin the index state, won't cabal produce a deterministic build plan anyway?

Reply:
i ve made it strict and trying the ghc-9.2 branch here: https://github.com/jneira/haskell-language-server/pull/59

Reply:
It seems it is working: https://github.com/jneira/haskell-language-server/runs/4714665789?check_suite_focus=true

Reply:
> SGTM.
> 
> Also, further dumb question: why do we need the freeze file? Given that we pin the index state, won't cabal produce a deterministic build plan anyway?

pr's can change dependencies removing, adding or updating bounds in any of the many .cabal files. This way we can have a dedicated cache taking in account those changes. We could hash all .cabal files though.

But otoh you can change things in the cabal.project (like constraints, flags,ghc option) which might affect dependencies. `cabal freeze` should record the relevant changes affecting dependencies. That way cosmetic or other changes not affecting dep resolution in cabal.project does not make miss the cache.

Reply:
Freeze file gets into account all projects & infers 1 package version set to rule them all.

Here `freeze` served the `caching` key detection purposes - so it was made fault-tolerant for the purpose of use.

There can be mutually exclusive dependencies, especially between `plugins`. Currently, most of them are enabled by default, so their package constraints should be solvable. But with the growing number of plugins - keeping all plugins package constraints up to date & mutually resolvable would be a challenge.

`freeze` in itself is a test, that all regular `build` targets in all projects can infer 1 set of versions of dependencies, so that is a dependency requirement test in itself & it probably should be made & placed so accordingly. Maybe now thinking on having CI to check with `freeze` for merge would seem viable.

And the additional thought - it seems a good strategy to check contributor PRs for consistency in update/dependency management. It would enforce also to have some backward compatibility of versions between projects, wich is both a good thing and may be a security concern, but security concern is when mainteiners can come into play. So for example, having the consistency means restricting from old versions in one place, would mean it can be at once removed in all other projects.

So indeed, having a `freeze` requirement seems as a feature.

Reply:
@michaelpj
> Also, further dumb question: why do we need the freeze file? Given that we pin the index state, won't cabal produce a deterministic build plan anyway?

It is not a dumb question. It is a Cabal innerworks obscurities. `index-state` determines the Hackage deps API catalog, version of the dependency catalog available at that time. Hackage allows a lot of versions for every package at every moment, so the catalog is a tree structure `package/{set-of-all-versions}`. So under one `index-state` two build targets are free choose two mutually exclusive versions.
`freeze` requires all (`build`, *afaik*) package descriptions to be solvable as a whole.